File: sdk/typescript/src/forecast/ForecastWorker.ts
// ===========================================================================
// @dynaep/core - Forecast Worker
// Async background worker that batches TimesFM inference calls and updates
// the ForecastCache. Coordinate events are buffered per-element, and the
// inference loop runs on a configurable interval, selecting pending and
// stale elements for batch prediction.
// ===========================================================================

import type { ForecastCache, CachedPrediction } from "./ForecastCache";
import type { BridgeClock } from "../temporal/clock";
import type {
  ForecastConfig,
  RuntimeCoordinates,
  RuntimeCoordinateEvent,
  ForecastPoint,
} from "../temporal/forecast";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/**
 * Client interface for batched TimesFM inference. Implementations may
 * target a local subprocess, a remote HTTP endpoint, or a mock for testing.
 */
export interface TimesFMClient {
  batchPredict(
    requests: BatchPredictRequest[]
  ): Promise<BatchPredictResponse[]>;
  available(): Promise<boolean>;
}

/**
 * A single element's prediction request within a batch.
 */
export interface BatchPredictRequest {
  elementId: string;
  series: Record<string, number[]>;
  horizon: number;
  contextLength: number;
}

/**
 * A single element's prediction response from the TimesFM backend.
 * Each dimension key maps to point estimates and quantile bounds.
 */
export interface BatchPredictResponse {
  elementId: string;
  predictions: Record<
    string,
    { point: number[]; quantile_low: number[]; quantile_high: number[] }
  >;
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/**
 * Numeric dimension keys extracted from RuntimeCoordinates for time-series
 * analysis. Non-numeric fields (visible, renderedAt) are excluded.
 */
const NUMERIC_DIMS = ["x", "y", "width", "height"] as const;

/** Minimum adaptive debounce in milliseconds. */
const MIN_DEBOUNCE_MS = 50;

/** Maximum adaptive debounce in milliseconds. */
const MAX_DEBOUNCE_MS = 2000;

/** Maximum number of elements to include in a single batch request. */
const MAX_BATCH_SIZE = 50;

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Clamp a number between a lower and upper bound.
 */
function clamp(value: number, lower: number, upper: number): number {
  return Math.max(lower, Math.min(upper, value));
}

// ---------------------------------------------------------------------------
// ForecastWorker
// ---------------------------------------------------------------------------

/**
 * Async background worker that batches TimesFM inference calls and keeps
 * the ForecastCache up to date. Coordinate events are buffered per-element,
 * and the inference loop selects pending and stale cache entries for
 * efficient batch prediction.
 */
export class ForecastWorker {
  private readonly cache: ForecastCache;
  private readonly timesfm: TimesFMClient;
  private readonly config: ForecastConfig;
  private readonly bridgeClock: BridgeClock;

  private coordinateBuffers: Map<string, RuntimeCoordinates[]>;
  private intervalHandle: ReturnType<typeof setInterval> | null;
  private running: boolean;

  constructor(
    cache: ForecastCache,
    timesfm: TimesFMClient,
    config: ForecastConfig,
    bridgeClock: BridgeClock
  ) {
    this.cache = cache;
    this.timesfm = timesfm;
    this.config = config;
    this.bridgeClock = bridgeClock;

    this.coordinateBuffers = new Map();
    this.intervalHandle = null;
    this.running = false;
  }

  // -------------------------------------------------------------------------
  // Lifecycle
  // -------------------------------------------------------------------------

  /**
   * Begin the async inference loop. The worker ticks at `config.debounceMs`
   * intervals, batching pending elements and dispatching them to the
   * TimesFM backend. Calling start() when already running is a no-op.
   */
  start(): void {
    if (this.running) {
      return;
    }

    this.running = true;
    this.intervalHandle = setInterval(() => {
      void this.tick();
    }, this.config.debounceMs);
  }

  /**
   * Stop the inference loop and release the interval timer. The worker
   * can be restarted by calling start() again. Coordinate buffers are
   * preserved across stop/start cycles.
   */
  stop(): void {
    if (this.intervalHandle !== null) {
      clearInterval(this.intervalHandle);
      this.intervalHandle = null;
    }

    this.running = false;
  }

  // -------------------------------------------------------------------------
  // Ingestion
  // -------------------------------------------------------------------------

  /**
   * Record an incoming runtime coordinate event. Called by the bridge when
   * `AEP_RUNTIME_COORDINATES` events arrive. The coordinates are appended
   * to the per-element buffer, trimmed to `config.contextWindow`, and the
   * element is marked as pending in the cache if no valid prediction exists.
   *
   * @param event - The runtime coordinate event to ingest.
   */
  recordCoordinates(event: RuntimeCoordinateEvent): void {
    const elementId = event.target_id;
    const coords: RuntimeCoordinates = {
      x: event.coordinates.x,
      y: event.coordinates.y,
      width: event.coordinates.width,
      height: event.coordinates.height,
      visible: event.coordinates.visible,
      renderedAt: event.coordinates.renderedAt,
    };

    let buffer = this.coordinateBuffers.get(elementId);
    if (!buffer) {
      buffer = [];
      this.coordinateBuffers.set(elementId, buffer);
    }

    buffer.push(coords);

    // Trim buffer to context window
    if (buffer.length > this.config.contextWindow) {
      const excess = buffer.length - this.config.contextWindow;
      buffer.splice(0, excess);
    }

    // Mark as pending if the cache has no valid prediction for this element
    const existing = this.cache.getCachedPrediction(elementId);
    if (!existing || existing.expiresAt <= this.bridgeClock.now()) {
      this.cache.markPending(elementId);
    }
  }

  // -------------------------------------------------------------------------
  // Inference Loop
  // -------------------------------------------------------------------------

  /**
   * Core inference loop tick. Selects pending and stale elements, builds
   * a batched prediction request, dispatches to TimesFM, and updates
   * the cache with results and adaptive debounce values.
   */
  private async tick(): Promise<void> {
    if (!this.running) {
      return;
    }

    const isAvailable = await this.timesfm.available();
    if (!isAvailable) {
      return;
    }

    // Collect elements that need prediction
    const selectedElements = this.selectElements();
    if (selectedElements.length === 0) {
      return;
    }

    // Build batch request
    const requests: BatchPredictRequest[] = [];
    for (const elementId of selectedElements) {
      const buffer = this.coordinateBuffers.get(elementId);
      if (!buffer || buffer.length < 3) {
        continue;
      }

      const series: Record<string, number[]> = {};
      for (const dim of NUMERIC_DIMS) {
        series[dim] = buffer.map((c) => c[dim]);
      }

      requests.push({
        elementId,
        series,
        horizon: this.config.forecastHorizon,
        contextLength: this.config.contextWindow,
      });
    }

    if (requests.length === 0) {
      return;
    }

    // Dispatch batch prediction
    let responses: BatchPredictResponse[];
    try {
      responses = await this.timesfm.batchPredict(requests);
    } catch (err: unknown) {
      const message =
        err instanceof Error ? err.message : "unknown error";
      console.warn(
        `[ForecastWorker] Batch prediction failed: ${message}`
      );
      return;
    }

    // Process responses
    for (const response of responses) {
      const prediction = this.parseResponse(response.elementId, response);
      this.cache.updateCache(response.elementId, prediction);

      const adaptiveDebounce = this.computeAdaptiveDebounce(
        response.elementId
      );
      this.cache.updateDebounce(response.elementId, adaptiveDebounce);
    }
  }

  /**
   * Select elements for the next batch prediction. Prioritizes pending
   * elements (cache misses), then fills remaining capacity with elements
   * whose predictions are oldest, up to the configured batch size.
   */
  private selectElements(): string[] {
    const batchSize = Math.min(
      MAX_BATCH_SIZE,
      this.config.maxTrackedElements
    );
    const selected: string[] = [];
    const selectedSet = new Set<string>();

    // Priority 1: pending elements (cache misses)
    const pending = this.cache.getPendingElements();
    for (const elementId of pending) {
      if (selected.length >= batchSize) {
        break;
      }
      if (this.coordinateBuffers.has(elementId)) {
        selected.push(elementId);
        selectedSet.add(elementId);
      }
    }

    // Priority 2: elements with coordinate buffers not yet in pending
    if (selected.length < batchSize) {
      for (const elementId of this.coordinateBuffers.keys()) {
        if (selected.length >= batchSize) {
          break;
        }
        if (selectedSet.has(elementId)) {
          continue;
        }
        const cached = this.cache.getCachedPrediction(elementId);
        if (!cached || cached.expiresAt <= this.bridgeClock.now()) {
          selected.push(elementId);
          selectedSet.add(elementId);
        }
      }
    }

    return selected;
  }

  // -------------------------------------------------------------------------
  // Response Parsing
  // -------------------------------------------------------------------------

  /**
   * Parse a BatchPredictResponse into a CachedPrediction. Extracts per-
   * dimension point estimates and quantile bounds, constructs ForecastPoint
   * arrays, and sets temporal metadata using the bridge clock.
   */
  private parseResponse(
    elementId: string,
    response: BatchPredictResponse
  ): CachedPrediction {
    const dimData: Record<
      string,
      { point: number[]; low: number[]; high: number[] }
    > = {};

    // Extract per-dimension arrays from the response
    for (const dim of NUMERIC_DIMS) {
      const entry = response.predictions[dim];
      if (entry) {
        dimData[dim] = {
          point: Array.isArray(entry.point) ? entry.point : [],
          low: Array.isArray(entry.quantile_low) ? entry.quantile_low : [],
          high: Array.isArray(entry.quantile_high)
            ? entry.quantile_high
            : [],
        };
      }
    }

    // Determine the number of forecast steps from the longest array
    let steps = 0;
    for (const dimKey of Object.keys(dimData)) {
      const len = dimData[dimKey].point.length;
      if (len > steps) {
        steps = len;
      }
    }

    // Build ForecastPoints for each step
    const points: ForecastPoint[] = [];
    const stepDurationMs =
      this.config.forecastHorizon / Math.max(steps, 1);

    for (let i = 0; i < steps; i++) {
      const predictedState: Partial<RuntimeCoordinates> = {};
      const quantileLow: Partial<RuntimeCoordinates> = {};
      const quantileHigh: Partial<RuntimeCoordinates> = {};

      for (const dim of NUMERIC_DIMS) {
        const entry = dimData[dim];
        if (entry) {
          predictedState[dim] = entry.point[i] ?? 0;
          quantileLow[dim] = entry.low[i] ?? entry.point[i] ?? 0;
          quantileHigh[dim] = entry.high[i] ?? entry.point[i] ?? 0;
        }
      }

      points.push({
        offsetMs: Math.round(stepDurationMs * (i + 1)),
        predictedState,
        quantileLow,
        quantileHigh,
      });
    }

    const forecastedAt = this.bridgeClock.now();
    const expiresAt = forecastedAt + this.config.forecastHorizon;

    return {
      targetId: elementId,
      predictions: points,
      confidence: 1.0,
      forecastedAt,
      expiresAt,
    };
  }

  // -------------------------------------------------------------------------
  // Adaptive Debounce
  // -------------------------------------------------------------------------

  /**
   * Compute an adaptive debounce interval for the given element based on
   * inter-event spacing in the coordinate buffer. Elements that receive
   * frequent updates get a shorter debounce; slowly-changing elements
   * get a longer one. The result is clamped to [50, 2000] milliseconds.
   *
   * When the buffer has fewer than 2 entries, the configured default
   * debounceMs is returned.
   */
  private computeAdaptiveDebounce(elementId: string): number {
    const buffer = this.coordinateBuffers.get(elementId);
    if (!buffer || buffer.length < 2) {
      return this.config.debounceMs;
    }

    // Compute inter-event intervals from renderedAt timestamps.
    // Each entry carries a renderedAt ISO string; we parse consecutive
    // pairs to derive real-time spacing.
    const intervals: number[] = [];
    for (let i = 1; i < buffer.length; i++) {
      const prevTime = Date.parse(buffer[i - 1].renderedAt);
      const currTime = Date.parse(buffer[i].renderedAt);

      // Only use valid, positive intervals
      if (
        !isNaN(prevTime) &&
        !isNaN(currTime) &&
        currTime > prevTime
      ) {
        intervals.push(currTime - prevTime);
      }
    }

    if (intervals.length === 0) {
      return this.config.debounceMs;
    }

    // Use the median inter-event interval as the adaptive debounce.
    // Median is robust against outliers from bursts or pauses.
    const sorted = [...intervals].sort((a, b) => a - b);
    const mid = Math.floor(sorted.length / 2);
    const medianInterval =
      sorted.length % 2 === 1
        ? sorted[mid]
        : (sorted[mid - 1] + sorted[mid]) / 2;

    return clamp(medianInterval, MIN_DEBOUNCE_MS, MAX_DEBOUNCE_MS);
  }
}

File: sdk/python/dynaep/forecast/forecast_worker.py
# ===========================================================================
# dynaep.forecast.forecast_worker - Async Background Worker
# Batched TimesFM inference worker using threading.Timer for periodic
# inference ticks. Collects coordinate buffers, dispatches batch predictions
# to the TimesFM client, parses responses into CachedPredictions, and
# updates the ForecastCache. Event processing is NEVER blocked.
# ===========================================================================

"""Async background worker for batched TimesFM inference."""
from __future__ import annotations

import math
import threading
import statistics
import logging
from dataclasses import dataclass, field
from typing import Optional, Protocol

from dynaep.temporal.clock import BridgeClock
from dynaep.temporal.forecast import ForecastConfig, RuntimeCoordinates, ForecastPoint
from dynaep.forecast.forecast_cache import ForecastCache, CachedPrediction

logger = logging.getLogger("dynaep.forecast.worker")

NUMERIC_DIMS = ("x", "y", "width", "height")


# ---------------------------------------------------------------------------
# TimesFM Client Protocol
# ---------------------------------------------------------------------------


class TimesFMClient(Protocol):
    """Protocol describing the interface to a TimesFM inference backend.

    Implementations may wrap a local TimesFM model, a remote HTTP endpoint,
    or a subprocess-based JSON-RPC bridge. The ForecastWorker only depends
    on this protocol, not on any concrete implementation.
    """

    def batch_predict(self, requests: list[dict]) -> list[dict]:
        """Run batched time-series predictions.

        Each request dict contains:
            - element_id (str): the target element identifier
            - series (dict[str, list[float]]): per-dimension numeric series
            - horizon (int): number of forecast steps
            - context_length (int): context window size

        Each response dict contains:
            - element_id (str): echoed back from the request
            - predictions (dict[str, dict]): per-dimension predictions with
              keys 'point', 'quantile_low', 'quantile_high' each mapping
              to a list of floats
            - confidence (float): model confidence score
        """
        ...

    def available(self) -> bool:
        """Return True if the backend is ready to accept predictions."""
        ...


# ---------------------------------------------------------------------------
# ForecastWorker
# ---------------------------------------------------------------------------


class ForecastWorker:
    """Background worker that periodically drains coordinate buffers,
    dispatches batched TimesFM inference, and updates the ForecastCache.

    Uses threading.Timer for the periodic inference loop rather than asyncio,
    matching the simpler threading model of the Python SDK. All buffer access
    is protected by a threading.Lock for thread safety.
    """

    def __init__(
        self,
        cache: ForecastCache,
        timesfm: TimesFMClient,
        config: ForecastConfig,
        bridge_clock: BridgeClock,
    ) -> None:
        self._cache = cache
        self._timesfm = timesfm
        self._config = config
        self._bridge_clock = bridge_clock

        self._coordinate_buffers: dict[str, list[RuntimeCoordinates]] = {}
        self._timer: Optional[threading.Timer] = None
        self._running: bool = False
        self._lock: threading.Lock = threading.Lock()

    # -----------------------------------------------------------------------
    # Lifecycle
    # -----------------------------------------------------------------------

    def start(self) -> None:
        """Begin the timer-based inference loop.

        Schedules the first tick using the configured debounce_ms interval.
        Subsequent ticks are rescheduled by _tick() after each inference
        cycle completes. If the worker is already running, this is a no-op.
        """
        if self._running:
            logger.debug("ForecastWorker already running, ignoring start()")
            return

        self._running = True
        interval_s = self._config.debounce_ms / 1000.0
        self._timer = threading.Timer(interval_s, self._tick)
        self._timer.daemon = True
        self._timer.start()
        logger.info(
            "ForecastWorker started with %.0fms tick interval",
            self._config.debounce_ms,
        )

    def stop(self) -> None:
        """Cancel the timer and stop the inference loop.

        Any in-flight _tick() call will complete but no further ticks
        will be scheduled. Coordinate buffers are preserved for potential
        restart.
        """
        self._running = False
        if self._timer is not None:
            self._timer.cancel()
            self._timer = None
        logger.info("ForecastWorker stopped")

    # -----------------------------------------------------------------------
    # Data Ingestion
    # -----------------------------------------------------------------------

    def record_coordinates(self, event: dict) -> None:
        """Record a coordinate event into the per-element buffer.

        Extracts target_id and coordinates from the event dict, appends
        the coordinates to the per-element ring buffer (trimmed to
        context_window), and marks the element as pending in the cache
        if no valid cached prediction exists.

        This method is thread-safe and non-blocking.

        Args:
            event: A dict containing at minimum 'target_id' and a
                'coordinates' dict with x, y, width, height, visible,
                and rendered_at fields. Falls back to 'element_id' if
                'target_id' is absent.
        """
        target_id = event.get("target_id") or event.get("element_id")
        if target_id is None:
            logger.debug("Event missing target_id/element_id, skipping")
            return

        coords_raw = event.get("coordinates", event)
        coords = RuntimeCoordinates(
            x=float(coords_raw.get("x", 0.0)),
            y=float(coords_raw.get("y", 0.0)),
            width=float(coords_raw.get("width", 0.0)),
            height=float(coords_raw.get("height", 0.0)),
            visible=bool(coords_raw.get("visible", True)),
            rendered_at=str(coords_raw.get("rendered_at", "")),
        )

        with self._lock:
            if target_id not in self._coordinate_buffers:
                self._coordinate_buffers[target_id] = []

            buffer = self._coordinate_buffers[target_id]
            buffer.append(coords)

            # Trim to context window
            if len(buffer) > self._config.context_window:
                excess = len(buffer) - self._config.context_window
                del buffer[:excess]

        # Mark pending in cache if no valid prediction exists
        cached = self._cache.get_cached_prediction(target_id)
        if cached is None:
            self._cache.mark_pending(target_id)

    # -----------------------------------------------------------------------
    # Inference Loop
    # -----------------------------------------------------------------------

    def _tick(self) -> None:
        """Core inference loop called by the timer.

        Drains pending elements from the cache, selects up to
        max_tracked_elements for batch inference, builds per-dimension
        series from coordinate buffers, dispatches to the TimesFM client,
        parses responses into CachedPredictions, updates the cache and
        adaptive debounce values, and reschedules the next tick.
        """
        try:
            self._run_inference_cycle()
        except Exception:
            logger.exception("ForecastWorker tick failed")
        finally:
            # Reschedule if still running
            if self._running:
                interval_s = self._config.debounce_ms / 1000.0
                self._timer = threading.Timer(interval_s, self._tick)
                self._timer.daemon = True
                self._timer.start()

    def _run_inference_cycle(self) -> None:
        """Execute a single inference cycle.

        Separated from _tick() for testability and clarity. This method
        performs the actual work of gathering pending elements, building
        batch requests, calling the TimesFM client, and updating the cache.
        """
        # Check backend availability
        if not self._timesfm.available():
            logger.debug("TimesFM backend unavailable, skipping inference cycle")
            return

        # Drain pending elements from cache
        pending = self._cache.get_pending_elements()
        if not pending:
            return

        # Limit to max_tracked_elements
        selected = pending[: self._config.max_tracked_elements]

        # Build batch requests from coordinate buffers
        batch_requests: list[dict] = []
        element_ids: list[str] = []

        with self._lock:
            for element_id in selected:
                buffer = self._coordinate_buffers.get(element_id)
                if buffer is None or len(buffer) < 3:
                    continue

                # Build per-dimension time series
                series: dict[str, list[float]] = {}
                for dim in NUMERIC_DIMS:
                    series[dim] = [getattr(c, dim) for c in buffer]

                batch_requests.append({
                    "element_id": element_id,
                    "series": series,
                    "horizon": self._config.forecast_horizon,
                    "context_length": self._config.context_window,
                })
                element_ids.append(element_id)

        if not batch_requests:
            return

        # Call TimesFM backend
        try:
            responses = self._timesfm.batch_predict(batch_requests)
        except Exception:
            logger.exception("TimesFM batch_predict failed")
            # Re-mark elements as pending so they are retried
            for eid in element_ids:
                self._cache.mark_pending(eid)
            return

        if responses is None:
            for eid in element_ids:
                self._cache.mark_pending(eid)
            return

        # Parse responses and update cache
        for i, response in enumerate(responses):
            if i >= len(element_ids):
                break

            resp_element_id = response.get("element_id", element_ids[i])
            try:
                prediction = self._parse_response(resp_element_id, response)
                self._cache.update_cache(resp_element_id, prediction)
            except Exception:
                logger.exception(
                    "Failed to parse response for element %s", resp_element_id
                )
                continue

            # Compute and update adaptive debounce
            debounce = self._compute_adaptive_debounce(resp_element_id)
            self._cache.update_debounce(resp_element_id, debounce)

    # -----------------------------------------------------------------------
    # Adaptive Debounce
    # -----------------------------------------------------------------------

    def _compute_adaptive_debounce(self, element_id: str) -> int:
        """Compute an adaptive debounce interval for the given element.

        Uses the length of the coordinate buffer as a proxy for update
        frequency: elements with dense histories get shorter debounce
        intervals (faster re-forecasting), while sparse histories get
        longer intervals. The result is clamped to [50, 2000] ms.

        Args:
            element_id: The element identifier to compute debounce for.

        Returns:
            The adaptive debounce interval in milliseconds.
        """
        with self._lock:
            buffer = self._coordinate_buffers.get(element_id)
            if buffer is None or len(buffer) < 2:
                return self._config.debounce_ms

            buffer_len = len(buffer)

        # Scale inversely with buffer density relative to context window.
        # A full buffer means high-frequency updates -> shorter debounce.
        # An almost-empty buffer means low-frequency updates -> longer debounce.
        fill_ratio = buffer_len / max(self._config.context_window, 1)

        # Interpolate between 2000ms (empty) and 50ms (full)
        raw_debounce = 2000.0 - (fill_ratio * (2000.0 - 50.0))

        clamped = max(50, min(2000, int(raw_debounce)))
        return clamped

    # -----------------------------------------------------------------------
    # Response Parsing
    # -----------------------------------------------------------------------

    def _parse_response(self, element_id: str, response: dict) -> CachedPrediction:
        """Parse a TimesFM response dict into a CachedPrediction.

        Expects the response to contain a 'predictions' dict keyed by
        dimension name (x, y, width, height), each with 'point',
        'quantile_low', and 'quantile_high' arrays. Constructs
        ForecastPoint objects for each forecast step.

        The forecasted_at timestamp is taken from the bridge-authoritative
        clock, and expires_at is set to forecasted_at + forecast_horizon.

        Args:
            element_id: The element this response corresponds to.
            response: The raw response dict from the TimesFM client.

        Returns:
            A CachedPrediction ready to be stored in the ForecastCache.
        """
        predictions_raw = response.get("predictions", {})

        # Extract per-dimension arrays
        dim_data: dict[str, dict[str, list[float]]] = {}
        for dim in NUMERIC_DIMS:
            entry = predictions_raw.get(dim)
            if entry is not None:
                dim_data[dim] = {
                    "point": entry.get("point", []) if isinstance(entry, dict) else [],
                    "low": entry.get("quantile_low", []) if isinstance(entry, dict) else [],
                    "high": entry.get("quantile_high", []) if isinstance(entry, dict) else [],
                }

        # Determine the number of forecast steps from the longest point array
        steps = 0
        for dim_entry in dim_data.values():
            point_len = len(dim_entry.get("point", []))
            if point_len > steps:
                steps = point_len

        # Build ForecastPoint objects
        predictions: list[ForecastPoint] = []

        if steps > 0:
            step_duration_ms = self._config.forecast_horizon / max(steps, 1)

            for i in range(steps):
                predicted_state: dict = {}
                quantile_low: dict = {}
                quantile_high: dict = {}

                for dim in NUMERIC_DIMS:
                    entry = dim_data.get(dim)
                    if entry is not None:
                        point_arr = entry.get("point", [])
                        low_arr = entry.get("low", [])
                        high_arr = entry.get("high", [])

                        predicted_state[dim] = (
                            point_arr[i] if i < len(point_arr) else 0.0
                        )
                        quantile_low[dim] = (
                            low_arr[i]
                            if i < len(low_arr)
                            else predicted_state.get(dim, 0.0)
                        )
                        quantile_high[dim] = (
                            high_arr[i]
                            if i < len(high_arr)
                            else predicted_state.get(dim, 0.0)
                        )

                predictions.append(
                    ForecastPoint(
                        offset_ms=round(step_duration_ms * (i + 1)),
                        predicted_state=predicted_state,
                        quantile_low=quantile_low,
                        quantile_high=quantile_high,
                    )
                )

        # Compute confidence from quantile spread
        confidence = self._compute_confidence(predictions)

        forecasted_at = self._bridge_clock.now()
        expires_at = forecasted_at + self._config.forecast_horizon

        return CachedPrediction(
            target_id=element_id,
            predictions=predictions,
            confidence=confidence,
            forecasted_at=forecasted_at,
            expires_at=expires_at,
        )

    # -----------------------------------------------------------------------
    # Confidence Estimation
    # -----------------------------------------------------------------------

    @staticmethod
    def _compute_confidence(predictions: list[ForecastPoint]) -> float:
        """Compute a confidence score from the quantile spread.

        Narrow quantile intervals indicate high confidence. The score is
        the average spread across all dimensions and all forecast steps,
        transformed via exponential decay to the range [0, 1].

        Args:
            predictions: The list of forecast points to score.

        Returns:
            A confidence value between 0.0 and 1.0.
        """
        if not predictions:
            return 0.0

        total_spread = 0.0
        count = 0

        for point in predictions:
            for dim in NUMERIC_DIMS:
                high = point.quantile_high.get(dim)
                low = point.quantile_low.get(dim)
                if isinstance(high, (int, float)) and isinstance(low, (int, float)):
                    total_spread += abs(high - low)
                    count += 1

        if count == 0:
            return 0.0

        avg_spread = total_spread / count
        decay_factor = 0.01
        confidence = math.exp(-decay_factor * avg_spread)
        return max(0.0, min(1.0, confidence))
